NonOverlappingCodes/2005-Markov-Clustering/mcl-14-137/doc/clminfo.1
.\" Copyright (c) 2014 Stijn van Dongen
.TH "clm info" 1 "16 May 2014" "clm info 14-137" "USER COMMANDS "
.po 2m
.de ZI
.\" Zoem Indent/Itemize macro I.
.br
'in +\\$1
.nr xa 0
.nr xa -\\$1
.nr xb \\$1
.nr xb -\\w'\\$2'
\h'|\\n(xau'\\$2\h'\\n(xbu'\\
..
.de ZJ
.br
.\" Zoem Indent/Itemize macro II.
'in +\\$1
'in +\\$2
.nr xa 0
.nr xa -\\$2
.nr xa -\\w'\\$3'
.nr xb \\$2
\h'|\\n(xau'\\$3\h'\\n(xbu'\\
..
.if n .ll -2m
.am SH
.ie n .in 4m
.el .in 8m
..
.SH NAME
clm info \- compute performance measures for graphs and clusterings\&.

clminfo is not in actual fact a program\&. This manual
page documents the behaviour and options of the clm program when
invoked in mode \fIinfo\fP\&. The options \fB-h\fP, \fB--apropos\fP,
\fB--version\fP, \fB-set\fP, \fB--nop\fP are accessible
in all \fBclm\fP modes\&. They are described
in the \fBclm\fP manual page\&.
.SH SYNOPSIS

\fBclm info\fP [options] <graph file> <cluster file> <cluster file>*

\fBclm info\fP
\fB[-o\fP fname (\fIwrite to file \fBfname\fP\fP)\fB]\fP
\fB[-pi\fP f (\fIapply inflation beforehand\fP)\fB]\fP
\fB[-tf\fP spec (\fIapply tf-spec to input matrix\fP)\fB]\fP
\fB[-cl-tree\fP fname (\fIexpect file with nested clusterings\fP)\fB]\fP
\fB[-cat-max\fP num (\fIdo at most \fBnum\fP tree levels\fP)\fB]\fP
\fB[-cl-ceil\fP <num> (\fIskip clusters of size exceeding <num>\fP)\fB]\fP
\fB[--node-self-measures\fP (\fIdump measure for native cluster\fP)\fB]\fP
\fB[--node-all-measures\fP (\fIdump measure for incident cluster\fP)\fB]\fP
\fB[-h\fP (\fIprint synopsis, exit\fP)\fB]\fP
\fB[--apropos\fP (\fIprint synopsis, exit\fP)\fB]\fP
\fB[--version\fP (\fIprint version, exit\fP)\fB]\fP
<matrix file> <cluster file> <cluster file>*
.SH DESCRIPTION

\fBclm info\fP computes several numbers indicative for the efficiency with
with a clustering captures the edge mass of a given graph\&.
Use it in conjunction with \fBclm dist\fP to determine which clusterings
you accept\&. See the EXAMPLES section in \fBclm dist\fP
for an example of \fBclm dist\fP and \fBclm info\fP (and \fBclm meet\fP) usage\&.
Output can be generated for multiple clusterings at the same time\&.

The \fBefficiency\fP factor is described in [1] (see
the \fBREFERENCES\fP section)\&. It tries to balance the dual aims of
capturing a lot of edges or edge weights and keeping the cluster footprint
or area fraction small\&. The efficiency number has several appealing
mathematical properties, cf\&. [1]\&. It is related to, but not derivable from,
the second and third numbers, the \fImass fraction\fP and the
\fIarea fraction\fP\&.

The \fBmass fraction\fP is defined as follows\&.
Let \fBe\fP be an edge of the graph\&. The clustering \fIcaptures\fP \fBe\fP
if the two nodes associated with \fBe\fP are in the same cluster\&.
Now the mass fraction is the joint weight of all captured edges divided
by the joint weight of all edges in the input graph\&.

The \fBarea fraction\fP is roughly the sum of the
squares of all cluster sizes for all clusters in the clustering, divided by
the square of the number of nodes in the graph\&. It says \fIroughly\fP,
because the actual formula uses the quantity \fBN\fP*(\fBN-1\fP) wherever it
says square (of \fBN\fP) above\&. A low/high area fraction indicates a
fine-grained/coarse clustering\&.
.SH OPTIONS

.ZI 2m "\fB-o\fP fname (\fIoutput file name\fP)"
\&
.br
.in -2m

.ZI 2m "\fB-pi\fP f (\fIapply inflation beforehand\fP)"
\&
.br
Apply inflation to the graph matrix and compute the performance
measures for the result\&.
.in -2m

.ZI 2m "\fB-tf\fP <tf-spec> (\fItransform input matrix values\fP)"
\&
.br
shared_defopt{-tf}
.in -2m

.ZI 2m "\fB-cl-tree\fP fname (\fIexpect file with nested clusterings (cone format)\fP)"
\&
'in -2m
.ZI 2m "\fB-cl-ceil\fP <num> (\fIskip (nested) clusters of size exceeding <num>\fP)"
\&
'in -2m
'in +2m
\&
.br
The specified file should contain a hierarchy of nested
clusterings such as generated by \fBmclcm\fP\&. The output is then
in a special format, undocumented but easy to understand\&.
Its purpose is to help cherrypick a single clustering
from a tree, in conjunction with the slightly experimental
and undocumented program \fBmlmfifofum\fP\&.

The measure that is used is very slow to compute for large clusters, and
generally it will be outside any interesting range (i\&.e\&. it will be small)\&.
Use \fB-cl-ceil\fP to skip clusters exceeding the specified size \-
\fBclm info\fP will directly proceed to subclusters if they exist\&.
.in -2m

.ZI 2m "\fB-cat-max\fP num (\fIdo at most num levels\fP)"
\&
.br
This only has effect when used with \fB-cl-tree\fP\&.
\fBclm info\fP will start at the most fine-grained level, working upwards\&.
.in -2m

.ZI 2m "\fB--node-all-measures\fP (\fIdump node-wise criteria for all incident clusters\fP)"
\&
'in -2m
.ZI 2m "\fB--node-self-measures\fP (\fIdump node-wise criteria for native cluster\fP)"
\&
'in -2m
'in +2m
\&
.br
These options return a key-value based format, with the meaning of
the keys as follows\&.

.di ZV
.in 0
.nf \fC
nm    file name (redundant unless multiple cluster files are provided)
ni    node index
ci    cluster index
nn    number of neighbours of this node (constant for a give node)
nc    cluster size (constant for a given cluster)
ef    efficiency for this node/cluster combination
em    max-efficiency for this node/cluster combination
mf    mass fraction: percentage of edge weights for this node in this cluster
ma    total mass of edge weights for this node in this cluster
xn    number of neighbours of the node that are not in the cluster
xc    number of nodes in the cluster that are not a neighbour of the node
ns    number of neighbours of the node that are also in this cluster
ti    the maximum of the edge weights for neighbours of this node that are in this cluster
to    the maximum of the edge weights for neighbours of this node that are NOT in this cluster
al    (alien) 1 if the node is not native to the cluster, 0 if the node is native
.fi \fR
.in
.di
.ne \n(dnu
.nf \fC
.ZV
.fi \fR

.in -2m
.SH AUTHOR

Stijn van Dongen\&.
.SH SEE ALSO

\fBmclfamily(7)\fP for an overview of all the documentation
and the utilities in the mcl family\&.
.SH REFERENCES

[1] Stijn van Dongen\&. \fIPerformance criteria for graph clustering and Markov
cluster experiments\fP\&. Technical Report INS-R0012, National Research
Institute for Mathematics and Computer Science in the Netherlands,
Amsterdam, May 2000\&.
.br
http://www\&.cwi\&.nl/ftp/CWIreports/INS/INS-R0012\&.ps\&.Z


NonOverlappingCodes/2005-Markov-Clustering/mcl-14-137/src/shcl/clminfo.c
/*   (C) Copyright 2001, 2002, 2003, 2004, 2005, 2006, 2007 Stijn van Dongen
 *   (C) Copyright 2008, 2009, 2010, 2011, 2012 Stijn van Dongen
 *
 * This file is part of MCL.  You can redistribute and/or modify MCL under the
 * terms of the GNU General Public License; either version 3 of the License or
 * (at your option) any later version.  You should have received a copy of the
 * GPL along with MCL, in the file COPYING.
*/


/* TODO
 *    Perhaps pre-read clusterings in one go, both for stack and
 *    herd (the default) cases.
 *
 *    create a sane output format.
 *    Either s-expression based or line based.
*/


#include <string.h>
#include <stdio.h>

#include "clm.h"
#include "report.h"
#include "clminfo.h"

#include "impala/matrix.h"
#include "impala/vector.h"
#include "impala/io.h"
#include "impala/app.h"
#include "impala/iface.h"
#include "impala/compose.h"

#include "clew/claw.h"
#include "clew/scan.h"
#include "clew/clm.h"
#include "clew/cat.h"

#include "mcl/interpret.h"
#include "mcl/transform.h"

#include "util/io.h"
#include "util/err.h"
#include "util/types.h"
#include "util/opt.h"
#include "util/minmax.h"

static const char* me  =  "clminfo";


enum
{  MY_OPT_OUTPUT = CLM_DISP_UNUSED
,  MY_OPT_CLTREE
,  MY_OPT_CLCEIL
,  MY_OPT_NCLMAX
,  MY_OPT_ADAPT
,  MY_OPT_PI
,  MY_OPT_TF
,  MY_OPT_PERNODE
,  MY_OPT_PERPAIR
,  MY_OPT_LAX
}  ;


static mcxOptAnchor infoOptions[] =
{  {  "-o"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_OUTPUT
   ,  "<fname>"
   ,  "output file name"
   }
,  {  "--node-self-measures"
   ,  MCX_OPT_DEFAULT
   ,  MY_OPT_PERNODE
   ,  NULL
   ,  "dump node-wise criteria for native cluster"
   }
,  {  "--node-all-measures"
   ,  MCX_OPT_DEFAULT
   ,  MY_OPT_PERPAIR
   ,  NULL
   ,  "dump node-wise criteria for all incident clusters"
   }
,  {  "-pi"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_PI
   ,  "<num>"
   ,  "apply inflation with parameter <num> beforehand"
   }
,  {  "-tf"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_TF
   ,  "<tf-spec>"
   ,  "first apply tf-spec to matrix"
   }
,  {  "-cl-tree"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_CLTREE
   ,  "<fname>"
   ,  "assume mclcm-type hierarchical clusterings (stack format)"
   }
,  {  "-cl-ceil"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_CLCEIL
   ,  "<num>"
   ,  "skip clusters with <num> or more members (with --cl-stack)"
   }
,  {  "-cat-max"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_NCLMAX
   ,  "<num>"
   ,  "do at most <num> levels of the hierarchy"
   }
,  {  "--lax"
   ,  MCX_OPT_HIDDEN
   ,  MY_OPT_LAX
   ,  NULL
   ,  "allow arbitrary sets to be read in, drop partition requirement"
   }
,  {  NULL ,  0 ,  0 ,  NULL, NULL}
}  ;


static mcxIO*  xfout        =  (void*) -1;
static mcxIO *xfmx          =  (void*) -1;
static mcxIO* xfstack       =  (void*) -1;
static mclx *mx             =  (void*) -1;
static dim     clceil       =  -1;
static dim     n_cl_max     =  -1;
static double inflation     =  -1;
static mcxbool perpair      =  -1;
static mcxbool pernode      =  -1;
static mcxbool cone         =  -1;
static mcxbool lax          =  -1;
static mcxTing* tfting      = (void*) -1;


static mcxstatus infoInit
(  void
)
   {  xfout          =  mcxIOnew("-", "w")
   ;  xfmx           =  NULL
   ;  xfstack        =  NULL
   ;  mx             =  NULL
   ;  clceil         =  0
   ;  n_cl_max       =  0
   ;  inflation      =  0.0
   ;  tfting         =  NULL
   ;  perpair        =  FALSE
   ;  pernode        =  FALSE
   ;  cone           =  FALSE
   ;  lax            =  FALSE
   ;  return STATUS_OK
;  }


static mcxstatus infoArgHandle
(  int optid
,  const char* val
)
   {  switch(optid)
      {  case MY_OPT_OUTPUT
      :  mcxIOrenew(xfout, val, "w")
      ;  break
      ;

         case MY_OPT_NCLMAX
      :  n_cl_max = atoi(val)
      ;  break
      ;

#if 0
               /* this leads to crashes; clew/claw.c requires nodes
                * to be in a cluster.
               */
         case MY_OPT_LAX
      :  lax = TRUE
      ;  break
      ;
#endif

         case MY_OPT_CLCEIL
      :  clceil = atoi(val)
      ;  break
      ;

         case MY_OPT_CLTREE
      :  xfstack = mcxIOnew(val, "r")
      ;  break
      ;

         case MY_OPT_PERPAIR
      :  perpair = TRUE
      ;  break
      ;

         case MY_OPT_TF
      :  tfting = mcxTingNew(val)
      ;  break
      ;

         case MY_OPT_PERNODE
      :  pernode = TRUE
      ;  break
      ;

         case MY_OPT_PI
      :  inflation = atof (val)
      ;  break
      ;

         default
      :  return STATUS_FAIL
      ;
      }
      return STATUS_OK
;  }


void do_stack
(  mclx* mx
,  mcxIO* xfstack
,  mcxIO* xfout
,  dim clceil
,  dim n_cl_max
)
   {  mclxCat st
   ;  dim i
   ;  mcxstatus status

   ;  mclxCatInit(&st)

   ;  status = mclxCatRead
         (  xfstack, &st, n_cl_max, NULL, mx->dom_rows
         ,  MCLX_PRODUCE_DOMSTACK | MCLX_ENSURE_ROOT | MCLX_REQUIRE_PARTITION
         )

   ;  if (status)
      mcxDie(1, me, "error reading stack")
   ;  else if (!st.level)
      mcxExit(0)

   ;  for (i=1;i<st.n_level;i++)
      {  mclx* cl = st.level[i].mx
      ;  mclx* clchild = st.level[i-1].mx

      ;  clmXPerformance(mx, clchild, cl, xfout, clceil)
      ;  fprintf(xfout->fp, "===\n")
   ;  }
      clmXPerformance(mx, st.level[st.n_level-1].mx, NULL, xfout, clceil)
;  }



static mcxstatus infoMain
(  int                  argc
,  const char*          argv[]
)
   {  int a =  0
   ;  mcxTing* ginfo = mcxTingEmpty(NULL, 40)

   ;  mcxLogLevel =
      MCX_LOG_AGGR | MCX_LOG_MODULE | MCX_LOG_GAUGE | MCX_LOG_WARN

   ;  if(1)
      mclx_app_init(stderr)

   ;  if(1)
      mclxIOsetQMode("MCLXIOVERBOSITY", MCL_APP_VB_YES)

   ;  if
      (  (xfstack && 1 != argc)
      || (!xfstack && 1 >= argc)
      )
      mcxDie
      (  1
      ,  me
      ,  "need %smatrix file%s"
      ,  xfstack ? "(just a) " : ""
      ,  xfstack ? "" : " and cluster file(s)"
      )

   ;  mcxIOopen(xfout, EXIT_ON_FAIL)

   ;  xfmx  =  mcxIOnew(argv[a++], "r")
   ;  mx    =  mclxReadx(xfmx, EXIT_ON_FAIL, MCLX_REQUIRE_GRAPH)

   ;  if (tfting)
      {  mclgTF* tfar = mclgTFparse(NULL, tfting)
      ;  if (!tfar)
         mcxDie(1, me, "errors in tf-spec")
      ;  mclgTFexec(mx, tfar)
   ;  }

      if (inflation)
      {  mclxInflate(mx, inflation)
      ;  mcxTingPrintAfter(ginfo, "inflation=%.2f\n", (double) inflation)
   ;  }

      mclxAdjustLoops(mx, mclxLoopCBmax, NULL)

   ;  if (xfstack)
         do_stack(mx, xfstack, xfout, clceil, n_cl_max)
      ,  mcxExit(0)

   ;  mcxIOfree(&xfmx)

   ;  while(a < argc)
      {  mcxIO *xfcl
      ;  mclx *cl
      ;  dim j
      ;  mclxCat st
      ;  mcxstatus status

      ;  if (!strcmp(argv[a], "--"))
         {  a++
         ;  fputc('\n', xfout->fp)
         ;  continue
      ;  }

         xfcl =  mcxIOnew(argv[a], "r")

      ;  mclxCatInit(&st)

      ;  if
         (( status
         =  mclxCatRead
            (  xfcl, &st, n_cl_max, NULL, mx->dom_rows
            ,  MCLX_PRODUCE_DOMSTACK | (lax ? 0 : MCLX_REQUIRE_PARTITION)
            )
         ))
         mcxDie(1, me, "error reading stack")

      ;  for (j=0;j<st.n_level;j++)
         {  cl = st.level[j].mx

         ;  if (pernode)
            clmDumpNodeScores(xfcl->fn->str, mx, cl, CLM_NODE_SELF)

         ;  else if (perpair)
            clmDumpNodeScores(xfcl->fn->str, mx, cl, CLM_NODE_INCIDENT)

         ;  else
            {  clmGranularityTable tbl
            ;  clmPerformanceTable pftable
            ;  mcxTing* linfo = mcxTingNew(ginfo->str)
            ;  mcxTingPrintAfter(linfo, " source=%s", xfcl->fn->str)
            ;  if (st.n_level > 1)
               mcxTingPrintAfter(linfo, ":%03d", (int) (j+1))

            ;  clmPerformance(mx, cl, &pftable)
            ;  clmPerformancePrint(xfout->fp, linfo->str, &pftable)

            ;  fputc(' ', xfout->fp)
            ;  clmGranularity(cl, &tbl)
            ;  clmGranularityPrint(xfout->fp, NULL, &tbl)
            ;  fputc('\n', xfout->fp)
         ;  }
            if (a < argc-1 || j <st.n_level-1)
            fprintf(xfout->fp, "===\n")
      ;  }

         mcxIOfree(&xfcl)
      ;  /* todo: free stack */
      ;  a++
   ;  }

      mclxFree(&mx)
   ;  return STATUS_OK
;  }


mcxDispHook* mcxDispHookInfo
(  void
)
   {  static mcxDispHook infoEntry
   =  {  "info"
      ,  "info [options] <mx file> <cl file>+"
      ,  infoOptions
      ,  sizeof(infoOptions)/sizeof(mcxOptAnchor) - 1
      ,  infoArgHandle
      ,  infoInit
      ,  infoMain
      ,  1
      ,  -1
      ,  MCX_DISP_MANUAL
      }
   ;  return &infoEntry
;  }


